extensions/items/standalone.mjs

```diff
@@ -0,0 +1,333 @@
+      return;
+    }
+    fromPlayerItem[itemName].amount -= amount;
+    this.items.set(playerID, fromPlayerItem);
+
+    const toPlayerItem = this.items.get(toPlayerID);
+    if (!itemName in toPlayerItem) {
+      toPlayerItem[itemName] = { amount: 0 };
+    }
+    toPlayerItem[itemName] += amount;
+    this.items.set(toPlayerID, toPlayerItem);
+
+    /* Notify the client that a certain amount of items have been given */
+    try {
+      await this.helper.callS2cAPI('items', 'onReceiveItem', 5000, toPlayerID, playerID, itemName, amount);
+    }
+    catch (e) {
+      // ignore if recipient is offline
+    }
+  }
+
+  /**
+   * Use an item once.
+   * @return {object} partials - An object with the following type:
+   * itemName: string;
+   * amount: number;
+   */
+  async c2s_useItem(playerID, itemName, amount) {
+    if (!this.items.has(playerID)) {
+      console.log('Player not initialized');
+      return;
+    }
+    if (!itemName in this.itemInfo || !itemName in this.itemInstances) {
+      console.log("Item does not exist");
+      return;
+    }
+    if (!this.itemInfo[itemName].usable) {
+      console.log("Item is not usable");
+      return;
+    }
+
+    const item = this.items.get(playerID);
+    if (!itemName in item || item[itemName].amount < amount) {
+      console.log("Insufficient quantity");
+      return;
+    }
+    item[itemName].amount -= amount;
+    this.items.set(playerID, fromPlayerItem);
+
+    itemInstances[itemName].useItem(amount);
+
+    try {
+      await this.helper.callS2cAPI('items', 'onUseItem', 5000, playerID, itemName, amount);
+    }
+    catch (e) {
+      // ignore if player becomes offline for some reason
+    }
+  }
+
+  /**
+   * This function is called by a client to notify the server whenever he/she drops an item on the ground.
+   * Once the client calls, the server will consequently update its database.
+   * itemName: string;
+   * amount: number;
+   */
+  async c2s_dropItem(playerID, mapCoord, facing, itemName) {
+    /* player has not registered yet */
+    if (!this.items.has(playerID)) {
+      console.log('Player not initialized');
+      return;
+    }
+
+    /* update database */
+    let itemObj = this.items.get(playerID);
+    if (!itemName in itemObj) {
+      console.log('No such item');
+      return;
+    }
+    itemObj[itemName].amount -= 1;
+    this.items.set(playerID, itemObj);
+
+    /* update cell set */
+    /* TODO: calculate dropped coordinate. Currently, the dropped coordinate is the same as the player's map coordinate */
+    this.droppedItemCell.set(this.droppedItemIndex, { "x": mapCoord.x, "y": mapCoord.y, "w": 1, "h": 1 });
+    this.droppedItemInfo.set(this.droppedItemIndex, itemName);
+    this.droppedItemIndex++;
+    for (const mapName of this.helper.gameMap.getOriginalCellSetStartWith('').keys()) { // Get all available maps.
+      this.gameMap.updateDynamicCellSet(mapName, 'droppedItem' + itemName.toUpperCase(), Array.from(this.droppedItemCell.values()));
```

Comment:
this.droppedItemCell seem to contain cells for all dropped items, and they may not be the same type as the current item. This seemed a bit off.

Reply:
Sorry, I forgot to fix this line. We should only update cells that contains a specific item by sieving whether `this.droppedItemInfo.get(idx)` equals to `itemName`.